game-core/src/main/java/games/strategy/triplea/settings/SelectionComponentFactory.java

```diff
@@ -749,27 +768,67 @@ public JComponent getUiComponent() {
+        // Only save when value changed
+        if (usernameField
+            .getText()
+            .equals(usernameSetting.getValue().map(String::new).orElse(""))) {
+          return;
+        }
+        try {
+          BackgroundTaskRunner.runInBackground(
+              "Fetching Login Token...",
+              () -> {
+                final NodeBbTokenGenerator tokenGenerator = new NodeBbTokenGenerator(forumUrl);
+                final Optional<Integer> oldUserId = uidSetting.getValue();
+                final Optional<char[]> oldToken = tokenSetting.getValue();
+                if (!usernameField.getText().isEmpty()) {
+                  final TokenInfo tokenInfo =
+                      tokenGenerator.generateToken(
+                          usernameField.getText(),
+                          new String(passwordField.getPassword()),
+                          Strings.emptyToNull(otpField.getText()));
+                  context.setValue(uidSetting, tokenInfo.getUserId());
+
+                  context.setValue(tokenSetting, tokenInfo.getToken().toCharArray());
+
+                  context.setValue(usernameSetting, usernameField.getText().toCharArray());
+                  // TODO error reporting
+                } else {
+                  context.setValue(usernameSetting, null);
+                  context.setValue(uidSetting, null);
+                  context.setValue(tokenSetting, null);
+                }
+
+                oldUserId.ifPresent(
+                    userId ->
+                        oldToken.ifPresent(
+                            token -> tokenGenerator.revokeToken(new String(token), userId)));
+              });
+        } catch (final InterruptedException e) {
+          Thread.currentThread().interrupt();
+        }
+
+        /*final String username = usernameField.getText();
+        context.setValue(uidSetting, username.isEmpty() ? null : username.toCharArray());
         withSensitiveArray(
             passwordField::getPassword,
             password ->
                 context.setValue(
                     passwordSetting,
                     (password.length == 0) ? null : password,
-                    SaveContext.ValueSensitivity.SENSITIVE));
+                    SaveContext.ValueSensitivity.SENSITIVE));*/
       }
 
       @Override
       public void resetToDefault() {
-        usernameField.setText(credentialToString(usernameSetting::getDefaultValue));
-        passwordField.setText(credentialToString(passwordSetting::getDefaultValue));
+        usernameField.setText("");
+        passwordField.setText("");
       }
 
       @Override
       public void reset() {
-        usernameField.setText(credentialToString(usernameSetting::getValue));
-        passwordField.setText(credentialToString(passwordSetting::getValue));
+        usernameField.setText(usernameSetting.getValue().map(String::new).orElse(""));
```

Comment:
I don't think `username` should be a `char[]`, IMO it's just not sensitive enough. Does it make sense changing `username` to be a String value before, after or in this PR?

Reply:
Yes and no.
1. It would make sense changing it to a normal string, but for usernames that are actually stored already it would fill the username field with a cryptic string for the first time, so we'd have to change the name of the setting to something different.

Regardless I think we should offer a button that performs a "factory reset" on the preferences store so users have a way to clear out settings that might no longer be there at all.

Reply:
IF we change datatype of a client setting, we only need to change the name of the setting. Having something clear out old settings would be nice, but I think certainly nice and not required.